source/cellular_pkthandler.c

```diff
@@ -94,117 +90,104 @@ static CellularPktStatus_t _convertAndQueueRespPacket( CellularContext_t * pCont
+        pktStatus = CELLULAR_PKT_STATUS_FAILURE;
+        LogError( ( "_convertAndQueueRespPacket: Got a response when the Resp Q is full!!" ) );
     }
 
     return pktStatus;
 }
 
 /*-----------------------------------------------------------*/
 
-/**
- * @brief copy the URC log in the buffer to a heap memory and process it.
- */
 static CellularPktStatus_t _processUrcPacket( CellularContext_t * pContext,
                                               const char * pBuf )
 {
     CellularPktStatus_t pktStatus = CELLULAR_PKT_STATUS_OK;
+    bool inputWithPrefix = false;
+    char * pInputLine = NULL;
+    char * pSavePtr = NULL, * pTokenPtr = NULL;
     CellularATError_t atStatus = CELLULAR_AT_SUCCESS;
-    char * payload = NULL;
 
-    if( pBuf != NULL )
+    LogDebug( ( "Next URC token to parse [%s]", pInputLine ) );
+
+    /* pBuf is checked in _Cellular_HandlePacket. */
+    atStatus = Cellular_ATStrDup( &pInputLine, pBuf );
+
+    if( atStatus != CELLULAR_AT_SUCCESS )
     {
-        atStatus = Cellular_ATStrDup( &payload, pBuf );
+        /* Fail to allocate memory. */
+        pktStatus = CELLULAR_PKT_STATUS_FAILURE;
+    }
+    else
+    {
+        /* Check if prefix exist in the input string. The pInputLine is checked in Cellular_ATStrDup. */
+        ( void ) Cellular_ATIsPrefixPresent( pInputLine, &inputWithPrefix );
 
-        if( atStatus == CELLULAR_AT_SUCCESS )
+        if( inputWithPrefix == true )
         {
-            /* The payload is null terminated. */
-            pktStatus = urcParseToken( pContext, ( char * ) payload );
-            Platform_Free( payload );
+            /* Cellular_ATIsPrefixPresent check the prefix string is valid and start with
+             * leading char. ":" is also checked in Cellular_ATIsPrefixPresent. Remove
+             * the leading char and split the string into the following substrings :
+             * pInputLine = "+" pTokenPtr + ":" + pSavePtr. */
+            pSavePtr = pInputLine + 1;
+            pTokenPtr = strtok_r( pSavePtr, ":", &pSavePtr );
+
+            if( pTokenPtr == NULL )
+            {
+                LogError( ( "_Cellular_AtParse : input string error, start with \"+\" but no token %s", pInputLine ) );
+                pktStatus = CELLULAR_PKT_STATUS_BAD_REQUEST;
+            }
+            else
+            {
+                pktStatus = _atParseGetHandler( pContext, pTokenPtr, pSavePtr );
+            }
         }
         else
         {
-            pktStatus = CELLULAR_PKT_STATUS_FAILURE;
-            LogWarn( ( "Couldn't allocate memory of %u for urc", ( uint32_t ) strlen( pBuf ) ) );
+            /* This is the input without prefix case. Nothing need to be done for this case.
+             * There are some special cases. For example, "+URC" the string without  ":" should
+             * be regarded as URC without prefix. */
+            pktStatus = _atParseGetHandler( pContext, pInputLine, pInputLine );
```

Comment:
I suggest to move `pktStatus = _atParseGetHandler( pContext, pTokenPtr, pSavePtr );` out of the condition.
Like:
```
if( inputWithPrefix == true )
{
   // ...
   if( pTokenPtr == NULL )
   {
       // ...
   }
}

if( pktStatus == CELLULAR_PKT_STATUS_OK )
{
    pktStatus = _atParseGetHandler( pContext, pTokenPtr, pSavePtr );
}
```

Reply:
The purpose of this function calls is to clearly indicate that if there is no prefix with the URC pInputLine will be used to search the URC handler and the parameter is also the pInputLine.